src/dfm/cleaning/quality_filter.py

```diff
@@ -0,0 +1,523 @@
+"""
+Danish implementation of quality filter described in [1]
+
+Authors:
+    Kenneth C. Enevoldsen
+    Kasper Junge
+    Malte Højmark-Bertelsen
+
+References: 
+    [1] Rae, J. W., Borgeaud, S., Cai, T., Millican, K., Hoffmann, J., Song, F.,
+    Aslanides, J., Henderson, S., Ring, R., Young, S., Rutherford, E., Hennigan,
+    T., Menick, J., Cassirer, A., Powell, R., Driessche, G. van den, Hendricks,
+    L. A., Rauh, M., Huang, P.-S., … Irving, G. (2021).
+    Scaling Language Models: Methods, Analysis & Insights from Training Gopher.
+    https://arxiv.org/abs/2112.11446v2
+"""
+
+from collections import Counter
+from functools import partial
+from typing import Iterable, Optional, Set, Tuple, List
+
+import spacy
+from spacy.tokens import Doc
+
+from nltk.tokenize import RegexpTokenizer
+
+from .repetitious_filter import (
+    set_dynamic_ext,
+    duplicate_fraction_getter,
+    duplicate_chr_fraction_getter,
+    top_ngram_chr_fraction,
+    duplicate_n_gram_fraction,
+)
+
+
+class NLTKTokenizer:
+    """
+    The NLTK tokenizer is noticeably faster than the spacy tokenizer
+    though performs notably worse tokenization:
+
+    Speed Comparison on a text of 237 tokens:
+
+    .. code:
+
+        text multiplication: time taken to tokenize
+
+        NLTK
+        10: 0.004389047622680664
+        100: 0.018768787384033203
+        1000: 0.17305397987365723
+        10000: 1.4743471145629883
+        SPACY
+        10: 0.019192934036254883
+        100: 0.15356707572937012
+        1000: 1.7956039905548096
+        10000: 18.097763776779175
+
+    Note this removes newlines and does not keep the text intact. Therefore it sets
+    the extension: doc._.text to keep the original text.
+
+    Example:
+        >>> nlp = spacy.blank("en")
+        >>> nlp.tokenizer = NLTKTokenizer(nlp.vocab)
+        >>> doc = nlp("What's happened to me? he thought. It wasn't a dream.")
+    """
+
+    def __init__(self, vocab):
+        self.vocab = vocab
+        self.tokenizer = RegexpTokenizer(pattern=r"\w+|[^\w\s]+")
+
+        if not Doc.has_extension("text"):
+            Doc.set_extension("text", default=None)
+
+    def __call__(self, text):
```

Comment:
Removed the custom tokenizer - I added it because it was notably faster than the spacy tokenizer when I tested using parallization. Turns out that is what is really slow about spaCy is the doc serialization (because it is lossless, which otherwise is great). The solution is to parallelize the filter instead of the nlp.pipe, thus you never need to serialize the doc object. There is more on this discussion [here](https://github.com/explosion/spaCy/discussions/10306).